### PR TITLE
API for acronym user fetch

### PR DESCRIPTION
**Pull request by loneil**
# Description
https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-774

Add the implementation of the acronyms//users endpoint.
Permission required for an acronym to call it.

Can test in the API tester with
acronyms/WORG/users

Altered the Keycloak management classes to take a filter in for getting users, though this one just fetches all KC users on the backend and then filters based on acronym mapping.
Could call endpoint once per user but that would be much more inefficient.

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-163

File: app/src/components/keyCloakServiceClientMgr.js [hunk 1]
```diff
@@ -151,20 +151,16 @@ class KeyCloakServiceClientManager {
...
   }
 
-  async findUser(username) {
-    log.debug('KeyCloakServiceClientManager.findUser', username);
-    if (!username) {
-      log.error('KeyCloakServiceClientManager.findUser', 'No user provided.');
-      throw new Error('Cannot get a user in KeyCloak realm: username required.');
-    }
+  // Get all users based on search param, supply undefined as searchParams to get ALL users in realm
+  async findUsers(searchParams) {
```
Comment (TimCsaky) [inline app/src/components/keyCloakServiceClientMgr.js:155]: I like the use of parameters as this function could be handy for retrieving a bigger subset of users for example in an admin stats page

File: app/src/components/acronyms.js [hunk 2]
```diff
@@ -26,6 +27,64 @@ const acronyms = {
...
+          userAcronymDetails: userAcronymDetails,
+          user: {
+            userId: au.userAcronym.userId,
+            keycloakGuid: kcUsr.id,
+            username: kcUsr.username,
+            firstName: kcUsr.firstName,
+            lastName: kcUsr.lastName,
+            email: kcUsr.email
+          }
+        };
```
Comment (jujaga) [inline app/src/components/acronyms.js:77]: Could in-line userAcronymDetails as part of the return object statement and save a const declaration.

File: app/src/components/keyCloakServiceClientMgr.js (hunk 1 above)
Comment (loneil) [reply app/src/components/keyCloakServiceClientMgr.js:155]: Indeed. Find available query params here https://www.keycloak.org/docs-api/5.0/rest-api/index.html#_users_resource

File: app/src/components/keyCloakServiceClientMgr.js [hunk 3]
```diff
@@ -151,20 +151,16 @@ class KeyCloakServiceClientManager {
...
+    log.debug('KeyCloakServiceClientManager.findUser', `Params: ${JSON.stringify(searchParams)}`);
```
Comment (jujaga) [inline app/src/components/keyCloakServiceClientMgr.js:156]: Function is plural now

File: app/src/components/keyCloakServiceClientMgr.js [hunk 4]
```diff
@@ -151,20 +151,16 @@ class KeyCloakServiceClientManager {
...
-    const users = await this.svc.getUsers(username);
+    const users = await this.svc.getUsers(searchParams);
 
     if (users && users.length) {
-      // Can only be one user by identified username (idir or github or whatever)
-      return users[0];
+      return users;
     } else {
-      log.debug('KeyCloakServiceClientManager.findUser', `No user found for ${username}`);
+      log.debug('KeyCloakServiceClientManager.findUser', `No users found for ${JSON.stringify(searchParams)}`);
```
Comment (jujaga) [inline app/src/components/keyCloakServiceClientMgr.js:163]: Function is plural now

File: app/src/components/users.js [hunk 5]
```diff
@@ -1,9 +1,31 @@
...
 const utils = require('./utils');
 
 const users = {
+  /**
+  * @function getAllGetokUsers
+  * Returns all the users in the keycloak realm that the GETOK api is running on. Use the api's client to call realm api
+  * @returns {object[]} An array of user objects
+  */
+  getAllGetokUsers: async () => {
+    log.debug('getAllGetokUsers');
```
Comment (jujaga) [inline app/src/components/users.js:16]: Half tempted to say this should be on verbose level as it's only letting you know you entered the function call.

File: app/src/routes/v1/acronyms.js [hunk 6]
```diff
@@ -51,6 +51,28 @@ acronymsRouter.get('/:acronym/clients', [
   }
 });
 
+/** Returns users from KC for the supplied acronym*/
```
Comment (jujaga) [inline app/src/routes/v1/acronyms.js:54]: Space before */

File: app/tests/unit/components/acronyms.spec.js [hunk 7]
```diff
@@ -55,6 +56,110 @@ describe('getAcronym', () => {
...
+    const acr = 'XXX';
+
+    const result = await acronyms.getUsers(acr);
+
+    expect(spy).toHaveBeenCalledTimes(1);
+    expect(spy).toHaveBeenCalledWith(acr);
+    expect(usersSpy).toHaveBeenCalledTimes(1);
+
+    expect(result).toBeTruthy();
+    expect(result.length).toEqual(2);
```
Comment (jujaga) [inline app/tests/unit/components/acronyms.spec.js:126]: Jest has an equivalent helper function for length evaluation: `expect(result).toHaveLength(2);`

File: app/src/routes/v1/acronyms.js [hunk 8]
```diff
@@ -51,6 +51,28 @@ acronymsRouter.get('/:acronym/clients', [
...
+acronymsRouter.get('/:acronym/users', async (req, res) => {
+  // Check for required permissions. Can only fetch details for the acronyms you are associated with
+  const permissionErr = await permissionHelpers.checkAcronymPermission(req.kauth.grant.access_token.content.sub, req.params.acronym);
```
Comment (parc-jason) [inline app/src/routes/v1/acronyms.js:57]: are we doing this kind of logic elsewhere?
we should move into middleware.

File: app/tests/unit/components/fixtures/acronymUsers.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+      "userAcronymId": "c",
+      "owner": false,
+      "createdAt": "2020-03-30T04:33:59.375Z",
+      "updatedAt": "2020-03-30T04:33:59.375Z",
+      "deletedAt": null,
+      "acronymId": "ccbb8b92-4a64-4c32-abb4-8cf5f15d7843",
+      "userId": "3"
+    }
+  }
+]
```
Comment (jujaga) [inline app/tests/unit/components/fixtures/acronymUsers.json:38]: No need to change right now, but I'm half tempted to say given this data structure, we should consider having this be a generator function instead -> that way we can dynamically assign created, updated, and acronymId uuids on each unit test run, and test for equivalence per instance instead.

File: app/src/routes/v1/acronyms.js (hunk 8 above)
Comment (loneil) [reply app/src/routes/v1/acronyms.js:57]: Yeah it's done in a number of places, that's why I refactored it out to the permissionhelpers util file a while back quickly.
Always intended on having it middleware fxn if possible, you're right that's the place for it I think. We can look at that for sure with next API changes.

File: app/src/components/users.js (hunk 5 above)
Comment (loneil) [reply app/src/components/users.js:16]: Yeah it's a trace statement so changing to verbose

File: app/src/components/acronyms.js (hunk 2 above)
Comment (loneil) [reply app/src/components/acronyms.js:77]: Yeah good point, looks a little cleaner too

File: app/src/components/keyCloakServiceClientMgr.js (hunk 3 above)
Comment (loneil) [reply app/src/components/keyCloakServiceClientMgr.js:156]: Fixed

File: app/src/components/keyCloakServiceClientMgr.js (hunk 4 above)
Comment (loneil) [reply app/src/components/keyCloakServiceClientMgr.js:163]: Fixed

File: app/src/routes/v1/acronyms.js (hunk 6 above)
Comment (loneil) [reply app/src/routes/v1/acronyms.js:54]: Fixed

File: app/tests/unit/components/acronyms.spec.js (hunk 7 above)
Comment (loneil) [reply app/tests/unit/components/acronyms.spec.js:126]: Cool

File: app/tests/unit/components/fixtures/acronymUsers.json (hunk 9 above)
Comment (loneil) [reply app/tests/unit/components/fixtures/acronymUsers.json:38]: Yeah could be more useful like that when needing more dynamic test results

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-163

**Pull request merged**